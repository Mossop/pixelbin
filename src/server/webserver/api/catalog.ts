import AWS from "aws-sdk";
import fetch from "node-fetch";

import { Api, Create, ObjectModel, Patch, ResponseFor, AWSResult } from "../../../model";
import { isoDateTime, now, s3Config, s3Params, s3PublicUrl } from "../../../utils";
import { UserScopedConnection } from "../../database";
import { ensureAuthenticated, ensureAuthenticatedTransaction } from "../auth";
import { AppContext } from "../context";
import { buildResponseMedia } from "./media";

export const testStorage = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    config: Api.StorageTestRequest,
  ): Promise<Api.StorageTestResult> => {
    let resultCode = AWSResult.UploadFailure;
    let target = "pixelbin-storage-test";
    try {
      let s3 = new AWS.S3({
        accessKeyId: config.accessKeyId,
        secretAccessKey: config.secretAccessKey,
        ...s3Config(config),
      });

      let content = isoDateTime(now());

      /* eslint-disable @typescript-eslint/naming-convention */
      await s3.upload({
        ...s3Params(config, target),
        Body: content,
        ContentLength: content.length,
      }).promise();
      /* eslint-enable @typescript-eslint/naming-convention */

      resultCode = AWSResult.DownloadFailure;
      let data = await s3.getObject(s3Params(config, target)).promise();

      if (!data.Body) {
        throw new Error("GetObject returned no content.");
      }
      let decoder = new TextDecoder();
      // @ts-ignore: TypeScript sees the Node types here.
      let result = decoder.decode(data.Body);
      if (result != content) {
        throw new Error("GetObject returned incorrect data.");
      }

      resultCode = AWSResult.PreSignedFailure;

      let url = await s3.getSignedUrlPromise("getObject", s3Params(config, target));

      let response = await fetch(url);
      if (response.status != 200) {
        throw new Error(`Pre-signed URL returned a failure status code (${response.statusText}).`);
      }
      let body = await response.text();
      if (body != content) {
        throw new Error("Pre-signed URL returned incorrect data.");
      }

      let publicUrl = s3PublicUrl(config, target);
      if (publicUrl) {
        response = await fetch(publicUrl);
        if (response.status != 200) {
          throw new Error(`Public URL returned a failure status code (${response.statusText}).`);
        }
        body = await response.text();
        if (body != content) {
          throw new Error("Public URL returned incorrect data.");
        }
      }

      await s3.deleteObject(s3Params(config, target)).promise();

      resultCode = AWSResult.DeleteFailure;
    } catch (e) {
      return {
        result: resultCode,
        message: e.message ? e.message : String(e),
      };
    }

    return {
      result: AWSResult.Success,
      message: null,
    };
  },
);

export const createStorage = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Api.StorageCreateRequest,
  ): Promise<Api.Storage> => {
    let storage = await userDb.createStorage(data);
    let {
      accessKeyId,
      secretAccessKey,
      owner,
      ...rest
    } = storage;
    return rest;
  },
);

export const createCatalog = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Create<ObjectModel.Catalog>,
  ): Promise<Api.Catalog> => {
    return userDb.createCatalog(data);
  },
);

export const listCatalog = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Api.CatalogListRequest,
  ): Promise<ResponseFor<Api.Media>[]> => {
    let media = await userDb.listMediaInCatalog(data.id);
    return media.map(buildResponseMedia);
  },
);

export const createAlbum = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Create<Api.Album>,
  ): Promise<Api.Album> => {
    return userDb.createAlbum(data.catalog, data);
  },
);

export const editAlbum = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Patch<Api.Album>,
  ): Promise<Api.Album> => {
    return userDb.editAlbum(data.id, data);
  },
);

export const listAlbum = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Api.AlbumListRequest,
  ): Promise<ResponseFor<Api.Media>[]> => {
    let media = await userDb.listMediaInAlbum(data.id, data.recursive);
    return media.map(buildResponseMedia);
  },
);

export const createTag = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Create<Api.Tag>,
  ): Promise<Api.Tag> => {
    return userDb.createTag(data.catalog, data);
  },
);

export const editTag = ensureAuthenticated(
  async (ctx: AppContext, userDb: UserScopedConnection, data: Patch<Api.Tag>): Promise<Api.Tag> => {
    return userDb.editTag(data.id, data);
  },
);

export const findTag = ensureAuthenticatedTransaction(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Api.TagFindRequest,
  ): Promise<Api.Tag[]> => {
    let parent: string | null = null;
    let foundTags: Api.Tag[] = [];

    for (let tag of data.tags) {
      let newTag = await userDb.createTag(data.catalog, {
        parent,
        name: tag,
      });
      foundTags.push(newTag);

      parent = newTag.id;
    }

    return foundTags;
  },
);

export const createPerson = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Create<Api.Person>,
  ): Promise<Api.Person> => {
    return userDb.createPerson(data.catalog, data);
  },
);

export const editPerson = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    data: Patch<Api.Person>,
  ): Promise<Api.Person> => {
    return userDb.editPerson(data.id, data);
  },
);
