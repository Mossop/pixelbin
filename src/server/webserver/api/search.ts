import type { Query, Api, ResponseFor } from "../../../model";
import { isCompoundQuery } from "../../../model";
import { isDateTime, isoDateTime } from "../../../utils";
import type { UserScopedConnection } from "../../database";
import { ensureAuthenticated } from "../auth";
import type { AppContext } from "../context";
import { buildResponseMedia } from "./media";

export const searchMedia = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    search: Api.MediaSearchRequest,
  ): Promise<ResponseFor<Api.Media>[]> => {
    let media = await userDb.searchMedia(search.catalog, search.query);
    return media.map(buildResponseMedia);
  },
);

function queryResponse(query: Query): ResponseFor<Query> {
  if (isCompoundQuery(query)) {
    return {
      ...query,
      queries: query.queries.map(queryResponse),
    };
  }

  if (isDateTime(query.value)) {
    return {
      ...query,
      value: isoDateTime(query.value),
    };
  }

  return query as ResponseFor<Query>;
}

export const createSavedSearch = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    search: Api.Create<Api.SavedSearch>,
  ): Promise<ResponseFor<Api.SavedSearch>> => {
    let saved = await userDb.createSavedSearch(search);
    return {
      ...saved,
      query: queryResponse(saved.query),
    };
  },
);

export const editSavedSearch = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    search: Api.Patch<Api.SavedSearch>,
  ): Promise<ResponseFor<Api.SavedSearch>> => {
    let saved = await userDb.editSavedSearch(search.id, search);
    return {
      ...saved,
      query: queryResponse(saved.query),
    };
  },
);

export const deleteSavedSearch = ensureAuthenticated(
  async (
    ctx: AppContext,
    userDb: UserScopedConnection,
    searches: string[],
  ): Promise<void> => {
    await userDb.deleteSavedSearch(searches);
  },
);
