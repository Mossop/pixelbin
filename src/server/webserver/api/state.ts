import { Api } from "../../../model";
import { AppContext } from "../context";

export async function buildUser(ctx: AppContext): Promise<Api.User | null> {
  if (!ctx.user) {
    return null;
  }

  let userDb = ctx.dbConnection.forUser(ctx.user.email);

  return {
    ...ctx.user,
    catalogs: await userDb.listCatalogs(),
    people: await userDb.listPeople(),
    tags: await userDb.listTags(),
    albums: await userDb.listAlbums(),
  };
}

async function buildState(ctx: AppContext): Promise<Api.State> {
  return {
    user: await buildUser(ctx),
  };
}

export async function getState(ctx: AppContext): Promise<Api.State> {
  return buildState(ctx);
}

export async function login(ctx: AppContext, data: Api.LoginRequest): Promise<Api.State> {
  await ctx.login(data.email, data.password);

  return buildState(ctx);
}

export async function logout(ctx: AppContext): Promise<Api.State> {
  await ctx.logout();

  return buildState(ctx);
}

export async function signup(ctx: AppContext, data: Api.SignupRequest): Promise<Api.State> {
  await ctx.dbConnection.createUser(data);
  await ctx.login(data.email, data.password);

  return buildState(ctx);
}
