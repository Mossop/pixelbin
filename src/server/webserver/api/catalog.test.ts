import { ErrorCode, AWSResult, emptyMetadata } from "../../../model";
import { RelationType } from "../../../model/api";
import { expect, getStorageConfig } from "../../../test-helpers";
import { connection, insertTestData, testData } from "../../database/test-helpers";
import { Table } from "../../database/types";
import { buildTestApp, expectUserState, fromCatalogs, catalogs, storage } from "../test-helpers";
import { savedSearchIntoResponse } from "./state";

const agent = buildTestApp();

beforeEach(insertTestData);

async function testStorage(id: string): Promise<void> {
  let request = agent();

  let config = await getStorageConfig(id);
  if (!config) {
    console.warn("Skipping test due to missing secrets.");
    return;
  }

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let response = await request
    .post("/api/storage/test")
    .send(config)
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body).toEqual({
    result: AWSResult.Success,
    message: null,
  });
}

test("Test AWS storage", async (): Promise<void> => {
  return testStorage("aws");
}, 30000);

test("Test Minio storage", async (): Promise<void> => {
  return testStorage("minio");
}, 30000);

test("Test Backblaze storage", async (): Promise<void> => {
  return testStorage("b2");
}, 30000);

test("Test Bad storage", async (): Promise<void> => {
  let request = agent();

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let response = await request
    .post("/api/storage/test")
    .send({
      accessKeyId: "foo",
      secretAccessKey: "bar",
      endpoint: "http://nowhere.foo",
      bucket: "buckit",
      region: "us-east-1",
      path: null,
      publicUrl: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body.result).not.toBe(AWSResult.Success);
  expect(response.body).toMatchSnapshot();
});

test("Create storage", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body.user.storage).toEqual([]);

  response = await request
    .put("/api/storage/create")
    .send({
      name: "My storage",
      accessKeyId: "foo",
      secretAccessKey: "bar",
      endpoint: null,
      bucket: "buckit",
      region: "somewhere-1",
      path: null,
      publicUrl: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body).toEqual({
    id: expect.stringMatching(/^S:[a-zA-Z0-9]+/),
    name: "My storage",
    endpoint: null,
    bucket: "buckit",
    region: "somewhere-1",
    path: null,
    publicUrl: null,
  });
  let storageId = response.body.id;

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body.user.storage).toEqual([{
    id: storageId,
    name: "My storage",
    endpoint: null,
    bucket: "buckit",
    region: "somewhere-1",
    path: null,
    publicUrl: null,
  }]);
});

test("Create catalog", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .put("/api/catalog/create")
    .send({
      storage: "s1",
      name: "Bad user",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/catalog/create")
    .send({
      storage: "s1",
      name: "Wrong user",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  response = await request
    .put("/api/storage/create")
    .send({
      name: "My storage",
      accessKeyId: "foo",
      secretAccessKey: "bar",
      endpoint: null,
      bucket: "buckit",
      region: "my-region-001",
      path: null,
      publicUrl: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);
  let storageId = response.body.id;

  response = await request
    .put("/api/catalog/create")
    .send({
      storage: storageId,
      name: "Good user",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let newCatalog = response.body;
  expect(newCatalog).toEqual({
    id: expect.stringMatching(/^C:[a-zA-Z0-9]+/),
    storage: storageId,
    name: "Good user",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00.000Z",
    verified: true,
    storage: [{
      id: storageId,
      name: "My storage",
      endpoint: null,
      bucket: "buckit",
      region: "my-region-001",
      path: null,
      publicUrl: null,
    }],
    catalogs: [
      ...testData[Table.Catalog],
      newCatalog,
    ],
    albums: testData[Table.Album],
    people: testData[Table.Person],
    tags: testData[Table.Tag],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });

  let ourStore = response.body.user.storage[0].id;

  // Can add to an existing storage.
  await request
    .put("/api/catalog/create")
    .send({
      storage: ourStore,
      name: "Existing catalog",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  await request
    .post("/api/login")
    .send({
      email: "someone2@nowhere.com",
      password: "password2",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  // But not if the storage is not accessible by the user.
  await request
    .put("/api/catalog/create")
    .send({
      storage: "s3",
      name: "Inaccessible",
    })
    .expect("Content-Type", "application/json")
    .expect(404);
});

test("Edit catalog", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .patch("/api/catalog/edit")
    .send({
      id: "c1",
      name: "Bad user",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .patch("/api/catalog/edit")
    .send({
      id: "c1",
      name: "Updated",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let updated = response.body;
  expect(updated).toEqual({
    id: "c1",
    storage: "s1",
    name: "Updated",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  let expected = [...testData[Table.Catalog]];
  expected[0].name = "Updated";

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00.000Z",
    verified: true,
    storage: [],
    catalogs: expected,
    albums: testData[Table.Album],
    people: testData[Table.Person],
    tags: testData[Table.Tag],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });

  await request
    .post("/api/login")
    .send({
      email: "someone3@nowhere.com",
      password: "password3",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .patch("/api/catalog/edit")
    .send({
      id: "c1",
      name: "Bad edit",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00.000Z",
    verified: true,
    storage: [],
    catalogs: expected,
    albums: testData[Table.Album],
    people: testData[Table.Person],
    tags: testData[Table.Tag],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });
});

test("Create album", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .put("/api/album/create")
    .send({
      catalog: "c1",
      name: "Bad user",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/album/create")
    .send({
      catalog: "c1",
      name: "Good user",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let newAlbum = response.body;
  expect(newAlbum).toEqual({
    id: expect.stringMatching(/^A:[a-zA-Z0-9]+/),
    name: "Good user",
    parent: null,
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00.000Z",
    verified: true,
    storage: [],
    catalogs: testData[Table.Catalog],
    albums: [
      ...testData[Table.Album],
      newAlbum,
    ],
    people: testData[Table.Person],
    tags: testData[Table.Tag],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });

  await request
    .post("/api/login")
    .send({
      email: "someone3@nowhere.com",
      password: "password3",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/album/create")
    .send({
      catalog: "c1",
      name: "Bad catalog",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Catalog."),
    },
  });
});

test("Edit album", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .patch("/api/album/edit")
    .send({
      id: "a1",
      name: "Bad name",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone2@nowhere.com",
      password: "password2",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  // Not including the ID is a failure.
  response = await request
    .patch("/api/album/edit")
    .send({
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(400);

  expect(response.body).toEqual({
    code: ErrorCode.InvalidData,
    data: {
      message: expect.stringContaining("decoder failed at key \"id\""),
    },
  });

  // Can't update albums in unowned catalogs.
  response = await request
    .patch("/api/album/edit")
    .send({
      id: "a6",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Album."),
    },
  });

  response = await request
    .patch("/api/album/edit")
    .send({
      id: "a1",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let updatedAlbum = response.body;
  expect(updatedAlbum).toEqual({
    id: "a1",
    name: "New name",
    parent: null,
    catalog: "c1",
  });

  response = await request
    .patch("/api/album/edit")
    .send({
      id: "a3",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let updatedAlbum3 = response.body;
  expect(updatedAlbum3).toEqual({
    id: "a3",
    name: "Album 3",
    parent: null,
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  let albums = fromCatalogs("c1", testData[Table.Album]);
  albums[0] = updatedAlbum;
  albums[2] = updatedAlbum3;
  expectUserState(response.body, {
    email: "someone2@nowhere.com",
    fullname: "Someone 2",
    created: "2010-01-01T00:00:00Z",
    verified: true,
    storage: storage([testData[Table.Storage][0]]),
    catalogs: catalogs("c1", testData[Table.Catalog]),
    albums: albums,
    people: fromCatalogs("c1", testData[Table.Person]),
    tags: fromCatalogs("c1", testData[Table.Tag]),
    searches: fromCatalogs("c1", testData[Table.SavedSearch]).map(savedSearchIntoResponse),
  });
});

test("List album and catalog", async (): Promise<void> => {
  /* eslint-disable-next-line */
  let ids = items => items.map(item => item.id);

  let request = agent();
  let db = await connection;
  let user1Db = db.forUser("someone1@nowhere.com");

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let { id: id1 } = await user1Db.createMedia("c1", emptyMetadata);
  await user1Db.setMediaRelations(RelationType.Album, [id1], ["a1"]);

  let { id: id2 } = await user1Db.createMedia("c1", emptyMetadata);
  await user1Db.setMediaRelations(RelationType.Album, [id2], ["a3"]);

  let { id: id3 } = await user1Db.createMedia("c1", emptyMetadata);
  await user1Db.setMediaRelations(RelationType.Album, [id3], ["a1", "a3"]);

  let { id: id4 } = await user1Db.createMedia("c1", emptyMetadata);
  await user1Db.setMediaRelations(RelationType.Album, [id4], ["a2"]);

  let response = await request
    .get("/api/album/list")
    .query({
      id: "a1",
      recursive: false,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(ids(response.body)).toInclude([
    id1,
    id3,
  ]);

  response = await request
    .get("/api/album/list")
    .query({
      id: "a3",
      recursive: false,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(ids(response.body)).toInclude([
    id2,
    id3,
  ]);

  response = await request
    .get("/api/album/list")
    .query({
      id: "a1",
      recursive: true,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(ids(response.body)).toInclude([
    id1,
    id2,
    id3,
  ]);

  response = await request
    .get("/api/catalog/list")
    .query({
      id: "c1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(ids(response.body)).toInclude([
    id1,
    id2,
    id3,
    id4,
  ]);
});

test("Create Tag", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .put("/api/tag/create")
    .send({
      catalog: "c1",
      name: "Bad user",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/tag/create")
    .send({
      catalog: "c1",
      name: "Good user",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let newTag = response.body;
  expect(newTag).toEqual({
    id: expect.stringMatching(/^T:[a-zA-Z0-9]+/),
    name: "Good user",
    parent: null,
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00Z",
    verified: true,
    storage: [],
    catalogs: testData[Table.Catalog],
    albums: testData[Table.Album],
    people: testData[Table.Person],
    tags: [
      ...testData[Table.Tag],
      newTag,
    ],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });

  await request
    .post("/api/login")
    .send({
      email: "someone3@nowhere.com",
      password: "password3",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/tag/create")
    .send({
      catalog: "c1",
      name: "Bad catalog",
      parent: null,
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Catalog."),
    },
  });
});

test("Edit tag", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .patch("/api/tag/edit")
    .send({
      id: "c1",
      name: "Bad name",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone2@nowhere.com",
      password: "password2",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  // Not including the ID is a failure.
  response = await request
    .patch("/api/tag/edit")
    .send({
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(400);

  expect(response.body).toEqual({
    code: ErrorCode.InvalidData,
    data: {
      message: expect.stringContaining("decoder failed at key \"id\""),
    },
  });

  // Can't update tags in unowned catalogs.
  response = await request
    .patch("/api/tag/edit")
    .send({
      id: "t5",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Tag."),
    },
  });

  response = await request
    .patch("/api/tag/edit")
    .send({
      id: "t1",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let updatedTag = response.body;
  expect(updatedTag).toEqual({
    id: "t1",
    name: "New name",
    parent: null,
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  let tags = fromCatalogs("c1", testData[Table.Tag]);
  tags[0] = updatedTag;
  expectUserState(response.body, {
    email: "someone2@nowhere.com",
    fullname: "Someone 2",
    created: "2010-01-01T00:00:00Z",
    verified: true,
    storage: storage([testData[Table.Storage][0]]),
    catalogs: catalogs("c1", testData[Table.Catalog]),
    albums: fromCatalogs("c1", testData[Table.Album]),
    people: fromCatalogs("c1", testData[Table.Person]),
    tags: tags,
    searches: fromCatalogs("c1", testData[Table.SavedSearch]).map(savedSearchIntoResponse),
  });
});

test("Find Tag", async (): Promise<void> => {
  let request = agent();

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let response = await request
    .post("/api/tag/find")
    .send({
      catalog: "c1",
      tags: ["tag2", "Tag6"],
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body).toEqual([{
    id: "t2",
    parent: null,
    name: "tag2",
    catalog: "c1",
  }, {
    id: "t6",
    parent: "t2",
    name: "Tag6",
    catalog: "c1",
  }]);

  response = await request
    .post("/api/tag/find")
    .send({
      catalog: "c1",
      tags: ["tag2", "tag6", "newtag"],
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  expect(response.body).toEqual([{
    id: "t2",
    parent: null,
    name: "tag2",
    catalog: "c1",
  }, {
    id: "t6",
    parent: "t2",
    name: "tag6",
    catalog: "c1",
  }, {
    id: expect.stringMatching(/T:[a-zA-Z0-9]+/),
    parent: "t6",
    name: "newtag",
    catalog: "c1",
  }]);
});

test("Create Person", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .put("/api/person/create")
    .send({
      catalog: "c1",
      name: "Bad user",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone1@nowhere.com",
      password: "password1",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/person/create")
    .send({
      catalog: "c1",
      name: "Good user",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let newPerson = response.body;
  expect(newPerson).toEqual({
    id: expect.stringMatching(/^P:[a-zA-Z0-9]+/),
    name: "Good user",
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  expectUserState(response.body, {
    email: "someone1@nowhere.com",
    fullname: "Someone 1",
    created: "2020-01-01T00:00:00Z",
    verified: true,
    storage: [],
    catalogs: testData[Table.Catalog],
    albums: testData[Table.Album],
    people: [
      ...testData[Table.Person],
      newPerson,
    ],
    tags: testData[Table.Tag],
    searches: testData[Table.SavedSearch].map(savedSearchIntoResponse),
  });

  await request
    .post("/api/login")
    .send({
      email: "someone3@nowhere.com",
      password: "password3",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  response = await request
    .put("/api/person/create")
    .send({
      catalog: "c1",
      name: "Bad catalog",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Catalog."),
    },
  });
});

test("Edit person", async (): Promise<void> => {
  let request = agent();

  let response = await request
    .patch("/api/person/edit")
    .send({
      id: "c1",
      name: "Bad name",
    })
    .expect("Content-Type", "application/json")
    .expect(401);

  expect(response.body).toEqual({
    code: ErrorCode.NotLoggedIn,
  });

  await request
    .post("/api/login")
    .send({
      email: "someone2@nowhere.com",
      password: "password2",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  // Not including the ID is a failure.
  response = await request
    .patch("/api/person/edit")
    .send({
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(400);

  expect(response.body).toEqual({
    code: ErrorCode.InvalidData,
    data: {
      message: expect.stringContaining("decoder failed at key \"id\""),
    },
  });

  // Can't update people in unowned catalogs.
  response = await request
    .patch("/api/person/edit")
    .send({
      id: "p4",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(404);

  expect(response.body).toEqual({
    code: ErrorCode.NotFound,
    data: {
      message: expect.stringContaining("Unknown Person."),
    },
  });

  response = await request
    .patch("/api/person/edit")
    .send({
      id: "p1",
      name: "New name",
    })
    .expect("Content-Type", "application/json")
    .expect(200);

  let updatedPerson = response.body;
  expect(updatedPerson).toEqual({
    id: "p1",
    name: "New name",
    catalog: "c1",
  });

  response = await request
    .get("/api/state")
    .expect("Content-Type", "application/json")
    .expect(200);

  let people = fromCatalogs("c1", testData[Table.Person]);
  people[0] = updatedPerson;
  expectUserState(response.body, {
    email: "someone2@nowhere.com",
    fullname: "Someone 2",
    created: "2010-01-01T00:00:00Z",
    verified: true,
    storage: storage([testData[Table.Storage][0]]),
    catalogs: catalogs("c1", testData[Table.Catalog]),
    albums: fromCatalogs("c1", testData[Table.Album]),
    people: people,
    tags: fromCatalogs("c1", testData[Table.Tag]),
    searches: fromCatalogs("c1", testData[Table.SavedSearch]).map(savedSearchIntoResponse),
  });
});
