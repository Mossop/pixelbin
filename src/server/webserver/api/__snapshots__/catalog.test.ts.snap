// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Bad storage 1`] = `
Object {
  "message": "Inaccessible host: \`nowhere.foo'. This service may not be available in the \`us-east-1' region.",
  "result": "upload-failure",
}
`;
